### ISSUE TITLE
float compare checks still(again?) failing

### ISSUE THREAD

**Issue by renemilk** [wwu-numerik/dune-stuff#63]
The automatically generated epsilon is std::complex(0,0) and a likely culprit.

**Comment by renemilk** [comment]
Nope. It isn't actually std::complex(0,0) that's just QtCreator capping display unless variable is expanded. Viewing real and imag separately, they're at the expected x*e-15. So now some of the tests still fail for complex numbers, but the infrastructure around complex handling should be at least a little nicer.
Maybe @tobiasleibner  or @BarbaraV  are interested in debugging the tests further?

**Comment by tobiasleibner** [comment]
This is due to std::complex behaving more like a vector than a scalar. For example, in this line
``` c++
    EXPECT_TRUE(Stuff::Common::FloatCmp::eq< relativeWeak   >(one, eps_plus));
```
one is 1 + i\*0 and eps_plus something like 1 + 1e-15 + i\*1e-15. So the comparison for the imaginary part looks like 1e-15 < 1e-15\*1e-15 for the relativeWeak and 1e-15 < 1e-15\*0 for relativeStrong. For a 2-vector, one = [1 1] instead of [1 0], so this does not happen for vectors.

**Comment by ftalbrecht** [comment]
Nicely found! I considered this with Rene a bit already: currently there is a specialization of `VectorAbstraction` for `std::complex` which eased some of the implementation. The longer I thought about it, however, the lesser I liked the idea since

* all vectors for which `VectorAbstraction` is specialized need to provide `operator[]` (complex does not, of course)
* one now has that `is_vector< std::complex< ... > >` is true
* and complex should behave more like a scalar than like a vector (since we want to use it as the scalar in vectors and matrices)

So I would propose to not specialize the `VectorAbstraction` for complex and instead put some more effort into float_cmp. Since the natural comparison that mostly happens would be between a complex and a scalar
```c++
Dune::Stuff::FloatCmp::eq(std::complex< double >(1.0 + 0.0i), 1.0);
```
and not a vector
```c++
Dune::Stuff::FloatCmp::eq(std::complex< double >(1.0 + 0.0i), std::vector< double >({1., 0.}));
```
we require some additional logic for complex, anyway.

Opinions?

**Comment by renemilk** [comment]
I am very strongly against anybody putting any more effort into float compare that is not directed towards replacing the macro code generation for various compare functions. That's a debug hell. Also we've decided not three weeks ago to treat complex like a vector.  If the VectorAbstraction included a read access to its abstracted vector we would indeed gain complex compare with no additional work and specializations like those currently in place.

**Comment by ftalbrecht** [comment]
I think the issue of the macros in float cmp and the handling of complex are not that much related. It is true that we agreed to handle complex as a vector because we thought it would be the best way to go. However, we have good reasons not to do so any more (see above), because we did some additional testing, thinking and debugging.
I agree that we should add read access to `VectorAbstraction` analogously to `MatrixAbstraction` as we have discussed. But that would not help us at all to get mathematically meaningful float comparison of complex numbers with other numbers and I do not think we will get that without a specialization of FloatCmp for complex...

**Comment by renemilk** [comment]
It's very strongly related because any change in the decided behaviour will need to be tested and debugged. That macro setup currently blocks stepping thru code altogether. That combined with the 3-5 Levels of indirection present would makes that very hard. 

Assuming you meant
```c++
Dune::Stuff::FloatCmp::eq(std::complex< double >(1.0, 0.0), 1.0);
``` 
I don't see a mathematical valid comparison rule that doesn't involve using the complex magnitude. Is that what you would now want?

As for for the points you listed:
> one now has that `is_vector< std::complex< ... > >` is true

That's not really a reason, that's the consequence of the choice to treat complex like a vector.

> * and complex should behave more like a scalar than like a vector (since we want to use it as the scalar in vectors and matrices)

I don't think that's a reason either. The tests use complex fields for {Field,std}vector just fine now (8ee067b7f273a890a44a10e08038ba11b6875e1a). 

Additionally, I am now convinced that users don't expect a comparison of magnitudes when calling float compare for two complex numbers, but element-wise for two vectors. If one really needs to compare a magnitude with a float it's very straightforward to call 
```c++
Dune::Stuff::FloatCmp::eq(std::complex< double >(1.0, 0.0).abs(), 1.0);
```

**Comment by ftalbrecht** [comment]
Don't get me wrong: I do see your pain of debugging those tests and I am completely on your side regarding any possible simplification of the situation in FloatCmp. However, I still disagree on the other points:

* I completely agree that one does not expect comparison by magnitude but I do not propose it either! If I want to compare the complex number `1.0 + 0.0*i` and the real number `1.0`, I would expect them to be equal. Our comparison functions should honor that by comparing the real part of the complex number to the other number and the complex part to `0.0`. This more or less corresponds to treating the double as a complex and comparing those, which is perfectly fine (mathematical and implementation wise).
* Regarding `is_vector`: I am aware that this is a consequence of our earlier choices. But I do now believe that this is wrong! The abstractions for vectors and matrices are not only there for our convenience when implementing string conversions and float comparisons. They are also used to generically handle vectors and matrices. Thus, there might be places where functions are guarded with `is_vector` that perform some vector computations, say a dot product. If one of those vectors may be a `std::complex` that is just plain wrong.

**Comment by renemilk** [comment]
You're not actually disagreeing with treating complex as a vector for comparison though. Your first point is exactly what that means. Remember there's an implicit conversion from double to complex<double> that is already doing what you think we should do. So from my point of view you want the comparison to work like a vector, but keep is_vector<complex> false. I can live with that, even if it's inconsistent and will always require special cases for complex, even if we generalize vector element access through the abstraction. I only implemented it that way because we agreed on treating it a like a vector itfp!

The trouble I ran into while debugging, which I think you're only guessing at, is a pretty good example of complicated code making it difficult to impossible to see/follow which function traces are actually getting called for varying inputs.

**Comment by ftalbrecht** [comment]
I do not disagree with the current complex comparison at all, correct! But I also think that any implementation issues should be discussed independently of the question of whether we want to treat complex as a vector by specializing `VectorAbstraction`.

Regarding the latter, I do see potential problems in constructs like https://github.com/wwu-numerik/dune-stuff/blob/master/dune/stuff/common/float_cmp_internal.hh#L64, for instance, where we check for `!is_complex && is_vector`. While the original proposal of treating complex as vectors was probably even my idea, I think code like the above may lead to serious misunderstanding in the future (even just considering us understanding what we did and why we did it).

Regarding your harsh critique of the FloatCmp implementation: I do not fully follow your argument. The code is obviously possible to understand and debug (for instance for @tobiasleibner, who did not implement any of that), though perhaps at a price. But we wanted to have float comparison functions which work for a wide range of inputs and these generic constructs are often not that pretty in the context of DUNE.

That being said, I am more than happy to welcome any simplifications in that area...

**Comment by renemilk** [comment]
Lets just agree that we're removing the VectorAbstraction specialization again.

> Regarding your harsh critique of the FloatCmp implementation: I do not fully follow your argument. The code is obviously possible to understand and debug (for instance for @tobiasleibner, who did not implement any of that), though perhaps at a price.

Debugging, as in using a debugger, is made difficult because you cannot really set breakpoints or watches in macro code. Error messages in macro generated code are also exceedingly difficult to read. That isn't particular to DUNE or generic code. @tobiasleibner did not find an issue with the macro generated code or code  called from that generated code, he found an issue with my test data input!

**Comment by ftalbrecht** [comment]
Anyway, I guess I have distracted us long enough from the original issue, sorry for that! I did not want to critique any of our code base but merely point out some issues I see regarding complex/vectors. Since I have not experienced any of them myself I am also quite satisfied with leaving things as they are and reconsidering matters as soon as one of us hits the next wall.

Regarding the original issue, did anyone solve that one already? Without looking too much at the code I would think that the problem might be related with the creation of `one`?

**Comment by renemilk** [comment]
Yeah, I did. Mentioned 96316f0d500a66bc3dd6cd7b04e5dd0792ee247b earlier.

**Comment by ftalbrecht** [comment]
Ah, nice! Should we proceed with a new issue to discuss simplifications of `FloatCmp`? Though I do not really have any quick ideas there...
